eZ/Publish/Core/Persistence/Cache/Tags/TagGenerator.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\Core\Persistence\Cache\Tags;
+
+use eZ\Publish\Core\Base\Exceptions\InvalidArgumentException;
+
+/**
+ * @internal
+ */
+final class TagGenerator implements TagGeneratorInterface
+{
+    private const PLACEHOLDER = '-%s';
+
+    /** @var string */
+    private $prefix;
+
+    /** @var array<string,string> */
+    private $patterns;
+
+    public function __construct(string $prefix, array $patterns)
+    {
+        $this->prefix = $prefix;
+        $this->patterns = $patterns;
+    }
+
+    /**
+     * @throws \eZ\Publish\Core\Base\Exceptions\InvalidArgumentException
+     */
+    public function generate(string $patternName, array $values = [], bool $withPrefix = false): string
+    {
+        if (!isset($this->patterns[$patternName])) {
+            throw new InvalidArgumentException($patternName, 'Undefined tag pattern');
+        }
+
+        $pattern = $this->patterns[$patternName];
+
+        if (empty($values)) {
+            $tag = str_replace(self::PLACEHOLDER, '', $pattern);
+        } else {
+            $tag = vsprintf($pattern, $values);
+        }
```

Comment:
```suggestion
        $tag = vsprintf($pattern, $values);
```

If there is no values provided then there is nothing to replace.

Reply:
What I wanted to achieve by this condition was handling generating tags used within several scenarios. For instance, location tag might be used as follows:
`l-14`, `ez-l`, `ez-l-13`.
If we don't specify arguments we don't want to end up with obsolete `-` at the end, especially if the tag is generated as a prefix.